api/proto/teleport/legacy/types/types.proto

```diff
@@ -233,6 +233,12 @@ message DatabaseServerSpecV3 {
   DatabaseV3 Database = 12 [(gogoproto.jsontag) = "database,omitempty"];
   // ProxyIDs is a list of proxy IDs this server is expected to be connected to.
   repeated string ProxyIDs = 13 [(gogoproto.jsontag) = "proxy_ids,omitempty"];
+  // LocalTime stores information about the server local time.
+  google.protobuf.Timestamp LocalTime = 14 [
```

Comment:
Should we refer to it as local time?

Local time represents the time in the machine's timezone, whereas `google.protobuf.Timestamp` does not include timezone information and is always in UTC.

It's important to note that the `*ServerSpec*` is not always updated or sent by the server. Our heartbeat mechanism prioritizes the KeepAlive (TTL extension at the database level) instead of sending the entire `*ServerSpec*` along with the upsert operation. As a result, the local time will reflect the time of the last change in the `*ServerSpec*` message.

However, if you use the watcher mentioned above, you will receive updates to `*ServerSpec*` each time the TTL key is extended, and this must be taken into account.

My suggestion is to use a separate process for time reconciliation, either through the `InventoryControlStream` or a different API, rather than relying on heartbeats. Using heartbeats could potentially cause issues for teleport.

Reply:
Thanks @tigrato, will check this one

Reply:
Was prototyping with inventory stream, tried to add local time as part of the metadata then after testing realized that is also cached data and refreshed only after Teleport restart, and time check should be similar to ping request, so replacing this one

Reply:
This might be better clarified by @rosstimothy, but we can probably use the `Ping/Pong` or similar messages.

We only need the time difference if there is a drift that satisfies the notification criteria, we do not need to store it anywhere. 
In this case, I think we can use [any clock synchronization algorithm](https://en.wikipedia.org/wiki/Network_Time_Protocol#Clock_synchronization_algorithm) will satisfy our needs. We can "measure" the time drift between two clocks which is what we need.

Since we want to measure the time difference between the auth and the agent, the auth should periodically (eg every 24 hours) issue the time sync requests and measure the difference.

Reply:
I've added additional command handling for time reconciliation, was not sure about ping/pong extension of new logic, just didn't want to mix responsibilities. We discussed about possible solution with making request to some predefined list of NTP servers, but this doesn't cover the case of blocked outbound traffic on the auth node.

Reply:
I think using the `UpstreamInventoryPong`  and  `DownstreamInventoryPing` is a better fit given that all the logic is already present and we just need to add a new field to it and change the comment. This allow us to skip cases where auth supports this feature but client's don't - `XXX_unrecognized` is set

I don't think we need an ntp server for this. The issue happens when auth clock drifts from the server(s). If they are both wrong but with the same value, there won't be any issue accessing the servers. 

Sure it would be great to compare them against a real cock, but it's not a hard requirement to make teleport work correctly.

Reply:
> This allow us to skip cases where  auth supports this feature but client's don't

we anyway need to add `google.protobuf.Timestamp LocalTime` to pong response in such case, so both auth and node should have these updates

Reply:
Here is a rephrased version:

There is a difference between not recognizing a oneof field and not knowing that a particular message field exists. 

For example, if auth sends a clock request, it expects a response from the server. If the server does not recognize the message type, it will ignore it, as seen here: https://github.com/gravitational/teleport/pull/44489/files#diff-13c41552d21a59b707b926d991bc325421f323363b3c6e802085f4068ed8197bR495-R497

Because it ignores the message, it will not respond. However, if you use Ping/Pong, and the response timestamp is empty, it indicates that the server does not support the clock check and auth can safely continue